From e6461ad91d7e1e1a6e4edd0689521f48052cb8c7 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:40:08 -0400
Subject: [virt] kvm/svm: Fix CPL export via SS.DPL

Message-id: <1406040016-3289-214-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86292
O-Subject: [RHEL7 PATCH v2 213/221] KVM: SVM: Fix CPL export via SS.DPL
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jan Kiszka <jan.kiszka@siemens.com>

We import the CPL via SS.DPL since ae9fedc793. However, we fail to
export it this way so far. This caused spurious guest crashes, e.g. of
Linux when accessing the vmport from guest user space which triggered
register saving/restoring to/from host user space.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 33b458d276bbdbe28febac0742835002b9f4778d)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/svm.c b/arch/x86/kvm/svm.c
index 4904767..8058d58 100644
--- a/arch/x86/kvm/svm.c
+++ b/arch/x86/kvm/svm.c
@@ -1462,6 +1462,7 @@ static void svm_get_segment(struct kvm_vcpu *vcpu,
    */
   if (var->unusable)
    var->db = 0;
+  var->dpl = to_svm(vcpu)->vmcb->save.cpl;
   break;
  }
 }
-- 
1.7.1